### ISSUE TITLE
Sudden and unjustified 3x increase in active usage costs from the 9th of June onwards

### ISSUE THREAD

**Issue by hesperanca** [microsoft/azure-container-apps#799]
### This issue is a: (mark with an x)
- [] bug report -> please search issues before submitting
- [ ] documentation issue or request
- [x] regression (a behavior that used to work and stopped in a new release)

### Issue description
We have noticed a dramatic increase in costs (3x) associated with all of our ACAs in both East US and EU North regions (the only regions where we have ACAs ) from the 9th of June onwards. There was no changes in usage patterns or new deployments on those dates to justify the increases. Drilling down on the charts we can see that the increases were mainly due to an increase in the Standard VCPU and memory active usages. The ACA is idle 99% of the time. The costs before the 9th were mainly idle costs  for CPU and memory. On the 9th they switched to active costs for no apparent reason (there were no changes in the usage pattern of the ACA).

To illustrate this we have attached below screenshots of the increases in costs on a specific ACA together with the CPU, replica, memory and requests metrics. The screenshots show that there were no changes in the CPU, nbr of replicas and nbr of requests or memory usage that can justify the cost increases. This is shown in just one of our ACAs but the same has happened for the other 20 ACAs we have across 2 subscriptions.

We are an ISV and were about to release a new SAS application into beta but had to pull the plug until we can justify these changes so this had a dramatic effect on the project.

**Expected behavior** [What you expected to happen.]
Costs to remain the same given that there were no changes in usage patterns

**Actual behavior** [What actually happened.]
Costs increased by 3x on average

**Screenshots**    
Costs:     
![image](https://github.com/microsoft/azure-container-apps/assets/29897247/8c8ce39a-79bb-4b45-8a93-95ff8fd68fc3)

CPU usage:
![image](https://github.com/microsoft/azure-container-apps/assets/29897247/feefa23c-b97f-4a5c-8802-acf80717cfcf)

Replica count:
![image](https://github.com/microsoft/azure-container-apps/assets/29897247/75d578e8-1a91-4a11-a096-7645329f32ea)

Nbr requests:
![image](https://github.com/microsoft/azure-container-apps/assets/29897247/51354088-a283-4c7b-b4e9-496c2b44d508)

Memory:
![image](https://github.com/microsoft/azure-container-apps/assets/29897247/2f310cb4-1b3f-4fe7-8e2d-270753e27bf9)

**Comment by maskati** [comment]
I am seeing the same. According to the [Container Apps pricing page](https://azure.microsoft.com/en-us/pricing/details/container-apps):

> A replica is active when vCPU usage is above 0.01 cores or when data received is above 1,000 bytes per second

Assuming that Container Apps metering is based on the same information as shown in [metrics](https://learn.microsoft.com/en-us/azure/container-apps/metrics#available-metrics), my replicas are often below these thresholds but are now being metered as active instead of idle.

**Comment by hesperanca** [comment]
> I am seeing the same. According to the [Container Apps pricing page](https://azure.microsoft.com/en-us/pricing/details/container-apps):
> 
> > A replica is active when vCPU usage is above 0.01 cores or when data received is above 1,000 bytes per second
> 
> Assuming that Container Apps metering is based on the same information as shown in [metrics](https://learn.microsoft.com/en-us/azure/container-apps/metrics#available-metrics), my replicas are often below these thresholds but are now being metered as active instead of idle.

This has been a massive problem for us. We have about 20 container apps. The daily cost of each has increased between 3x and 5x so we had to shut down most of them to avoid costs spiralling out of control. I hope someone will look into this soon.

**Comment by torosent** [comment]
We are investigating. If there is a billing error then it will be corrected.

**Comment by torosent** [comment]
@hesperanca , Can you send us the app name and subscription id to `acasupport st microsoft dot com`?

**Comment by hesperanca** [comment]
> @hesperanca , @maskati , Can you send us the app name and subscription id to `acasupport at microsoft dot com`?

Hi @torosent. Just sent you the email with the details. Please let me know if you need any further info.

**Comment by stroborobo** [comment]
We're seeing the same thing.

I'm not sure if it's related at all, but I noticed that the contains report having 16G of physical memory. (/proc/meminfo MemTotal)

**Comment by stroborobo** [comment]
There were also a ton of restart events we didn't trigger ourselves from 2023-06-08 7:30 until 2023-06-10 0:00 UTC, roughly 12 times each hour.

**Comment by hesperanca** [comment]
> There were also a ton of restart events we didn't trigger ourselves from 2023-06-08 7:30 until 2023-06-10 0:00 UTC, roughly 12 times each hour.

Interesting. We have the same. It looks like the restarts were caused by the default TCP probes (they do not show on the probes list in the portal). There have been reports of a few network issues at the time so I wonder if the probes lost connection with the ACA and thought it was down so it decided to restart it.

**Comment by torosent** [comment]
Hi,

The increased active usage is not related to any of the mentioned issues. 
We will issue comms to all our customers explaining the increased usage.

For all other issues please open an issue in GH or send us an email to 'acasupport at microsoft dot com'

**Comment by torosent** [comment]
Hi,

This is still an active investigation. 

Tomer

**Comment by JennyLawrance** [comment]
**Update @ 6/21/2023 20:15 UTC:**
We have fully understood the root cause of the billing discrepancy and are working on a fix right now. 
Tentative ETA for fix rollout completion: 6/28/2023

**Comment by hesperanca** [comment]
@JennyLawrance Thanks for the update. I'm assuming there will be a credit back into our Azure accounts once things are fixed? 

This issue is burning our Azure credits at an alarming rate.

**Comment by JennyLawrance** [comment]
@hesperanca, Yes. We are figuring out the logistics of this as well in parallel. Will post an update when we have the specifics to share.

**Comment by JennyLawrance** [comment]
**Update @ 6/28/2023 17:18 UTC**

The Azure Container Apps product team has recently identified an issue where vCPU was being over-calculated relative to the published Azure Container Apps pricing which has resulted in overbilling for some apps hosted in Azure Container Apps. This issue began on June 9th, 2023 and has since been fixed as of June 24th, 2023. _For the curious, there was a missing divided by time interval bug in the code that checked if the container app crossed its idle CPU thresholds, thereby drastically overestimating the containers as active, even though they were idle at that time_.

We will be issuing refunds for overbilled amounts during the given time period (6/9-6/24) for impacted subscriptions. This refund will not appear on your June billing statement but will be provided as a credit in the next billing statement. If your subscription was impacted and will be reimbursed, you will receive an email from the team by early next week.

**Comment by hesperanca** [comment]
@JennyLawrance . Thanks for the update and fix. I'm assuming that this will also apply to those of us that have an Azure Sponsorship account (we had a big hammering in our credits)?

**Comment by hesperanca** [comment]
Thanks for the update @JennyLawrance

**Comment by JennyLawrance** [comment]
@hesperanca We have an update from the commerce team regarding refunds to sponsorship accounts. We learned that the sponsored accounts are excluded from automatic refunds, and you will need to request a refund manually by raising a service request via [Azure Portal](https://ms.portal.azure.com/#view/Microsoft_Azure_Support/HelpAndSupportBlade/~/overview)

**Comment by hesperanca** [comment]
@JennyLawrance Thanks. I'll chase this with the Azure support team. Thanks for your help in sorting this out.

**Comment by vinicius-batista** [comment]
Hi, we've also having this problem since 12th jun but we still seeing this increase.

Anyone still seeing this increase?

**Comment by dinoo** [comment]
We are seeing a similar issue with a 3x times increase in calculated usage costs starting from the 31st of October 2023 across all our azure container apps and onwards.

I have submitted a ticket to Microsoft via the Azure Portal.

@JennyLawrance is the issue with the billing pipeline back?

**Comment by rytmis** [comment]
Same pattern showing up in Cost Analysis since Oct. 31st (with a mystery gap on Nov 1st). Looks like all of vCPU/Memory Active/Idle have roughly doubled. The absolute costs aren't that alarming _yet_, because this is a small demo environment, but the pattern is somewhat alarming. Notably, we can't detect any changes to the resources, nor have the container images changed at the time of the cost change:

![image](https://github.com/microsoft/azure-container-apps/assets/121952/ef065810-d434-433e-ad65-e41428a558ce)

**Comment by JennyLawrance** [comment]
@dinoo, @rytmis can you send us your subscription id/container app name to acasupport(at)microsoft(dot)com? We will investigate.

**Comment by dinoo** [comment]
It seems that we were being underbilled for all our Azure Container Apps that are hosted in the _West Europe_ region.
More information about it can be found on [Issue:-West-Europe-Underbilling](https://github.com/microsoft/azure-container-apps/wiki/Issue:-West-Europe-Underbilling)

**Comment by hesperanca** [comment]
These billing issues with ACAs are a serious concern for us. We just built an entire multitenant SAAS solution based on ACAs and these random variations in price are a serious challenge for our pricing model.

**Comment by cachai2** [comment]
Thank you @dinoo and @hesperenca for both flagging that you had not received notification of the fix as well as sharing your concerns with the team. We understand the concerns around billing, and it is the team’s responsibility to ensure that billing is predictable for customers. We want to be fully transparent with what the team has been working on, why these billing changes have occurred, and how we plan to improve going forward, especially as we have had multiple occurrences of billing issues in the past year.

We are striving to build a reliable service. After the first billing issue occurred, we began an exhaustive investigation into our platform to solidify our billing pipeline (both in terms of better detection and exhaustive testing). This is how we identified the issue in West Europe that got fixed November 1st where billing meters weren’t emitting for the majority of the time. This was due to bottlenecks in our billing platform that couldn’t keep up with the number of customers in the region. We’ve proactively fixed this issue now for all other regions, but customers in West Europe were impacted. Upon further investigation, we found that the billing issue has been occurring since May 2023 with increasing amounts of time that meters weren’t correctly emitted. We will not be looking to back charge for these underbilled amounts.

We are further strengthening our testing/regression suite for billing and adding additional monitoring to ensure that billing is behaving as expected. This work is currently in progress and is estimated to be completed by end of year. This will improve our ability to self-detect issues as soon as they happen and allow us to fix them right away to minimize any impact. 

We apologize for the inconvenience caused, and we are taking steps to improve. If you have additional thoughts and concerns, please do continue to share them as they are valuable to the team.

A final note: for this specific billing issue in West Europe, we had sent out communications to customers we believed to be impacted in October, 2023. However, it appears that some impacted customers didn’t receive the notification. Hence, for future comms, we will create a Github issue in addition to sending out direct notifications to ensure that there is public visibility for all.

Thank you,
The ACA team

**Comment by morgan-cromell** [comment]
We have around a 3x cost increase for our container app billing as of 1 November. We are unsure if we have been affected by the underbilling as the cost has been low before May 2023 and this was the first increase we have seen. We cannot go back to November 3 anymore to check as explained in https://github.com/microsoft/azure-container-apps/wiki/Issue:-West-Europe-Underbilling

**Comment by cachai2** [comment]
@morgan-cromell thanks for flagging this. We'll create a separate detector specifically for the underbilling time frame and update the wiki information. This work should be completed by early next week, and I'll update this thread.

**Comment by cachai2** [comment]
Hi @morgan-cromell, thanks for the patience. We've created a detector specifically for the billing issue which is hardcoded to the timeframe centered around 11/1/23. The detector is called 'West Europe Billing Issue.' Exact instructions have been updated in the wiki: https://github.com/microsoft/azure-container-apps/wiki/Issue:-West-Europe-Underbilling.

Please let me know if there are any additional issues.

**Comment by morgan-cromell** [comment]
@cachai2 This tool is not working for us. We get the following error:
`This detector is only for resources in West Europe, this resource is in: West Europe`

![image](https://github.com/microsoft/azure-container-apps/assets/93519083/e619aae2-9f6f-49a3-bf12-b63566bebad9)

**Comment by cachai2** [comment]
@morgan-cromell the logic should be set so that resources in West Europe will see the graphs. Resources not in West Europe will see the screenshot you shared. Can you verify the app you're checking the detector from is in West Europe? If it's in West Europe, we'll double check the detector.

If your resource that you saw the billing increase for isn't in West Europe, can you share more details about your app name, subscription, etc to acasupport at microsoft.com?

**Comment by morgan-cromell** [comment]
@cachai2 the resource is in west europe.  It says so in the screenshot and error message. We have also confirmed elsewhere that the resource is in west europe.

**Comment by maskati** [comment]
Same here. Resource in West Europe but detector claiming otherwise. Maybe related to #795?

**Comment by cachai2** [comment]
@morgan-cromell. Ah, thank you. Apologies, I missed that. We'll make a fix and see if it's related at all to the issue @maskati linked.

**Comment by tdaroly** [comment]
@morgan-cromell Can you provide the name of the resource that you are seeing this on ?

**Comment by tdaroly** [comment]
The reported bug has been successfully fixed and the solution has been deployed to production. If you have any questions, feel free to reach out.

**Comment by arnoldligtvoet** [comment]
We also seem to be effected by the 'under billing' issue. Costs x2.7 since November first without any sort of interaction from our side. No notification whatsoever from Microsoft Azure teams. 

Checking the 'West Europe Billing Issue' I do notice something weird. For a container app that has a min replica count of 2 and a max of 3 I get this graph:
![image](https://github.com/microsoft/azure-container-apps/assets/26329745/bdb777f3-0e3c-4593-b48b-99c75447617a)

It shows a minimum of 3 replica's and a max of 5? If I look at the metrics for the app for the same time frame it shows an average of 2.5, max of 3. 

It's also quite concerning that the billing/cost can change without notice and it's killing for our business cases using Azure Container Apps.

**Comment by Sebs5i** [comment]
We faced the same issue and even raised a support ticket, but it seems they have not been aware of this issue. 
Prices increased almost 4x and our monthly billing skyrocked with no upfront information form Microsoft.
Container are placed in west europe and i ran the detector. Just for my understanding. Because the idle time is up continiously instead of fluctuating we are now charged more, correct @cachai2 ? 
![image](https://github.com/microsoft/azure-container-apps/assets/56294127/4d7c5529-f7fc-40eb-89ec-d2c5a7018b69)

**Comment by cachai2** [comment]
@Sebs5i thanks for raising this. Previous to the fix at the start of November, significant portions of your apps' usage weren't being emitted. This is why you see the drops in usage. After the fix, usage is being emitted correctly, and we aren't back charging for any of the previously underbilled amounts. We sent notifications to customers in West Europe but have since been made aware that not all customers received the communication. I apologize that your subscriptions were missed.

For more information, see the note here: [West Europe Underbilling Issue](https://github.com/microsoft/azure-container-apps/issues/799#issuecomment-1819995124).

**Comment by hesperanca** [comment]
This is quite a big problem. Especially for ISVs that are just launching a SAAS product based on ACAs (like us). They have calculated their baseline infrastructure costs at 3x less than what it's actually going to cost them. If they charge their customers annually, they will probably make a loss on those accounts or have to have an awkward conversation with their customers. I think that, for those customers that have been affected, Microsoft should take this into consideration and honour the old pricing model for a certain period of time.

**Comment by cachai2** [comment]
Hi @hesperanca I hear your concerns, but the pricing model has not changed. We are billing for the amounts listed on our pricing page. The difference is that previously for a subset of customers they were being billed less than the advertised amounts and getting free compute. Estimates made using the prices on the pricing page will not be impacted. That being said, I'm happy to sync and understand more about your scenario as well as any impact it may have had on your products. Feel free to reach out to me on Linkedin, and we can get a call setup.

**Comment by cachai2** [comment]
Closing this thread as the original June 9th, 2023 billing issue has long since been resolved. For future billing issues, please open a new issue.
